keps/sig-storage/3314-csi-volume-snapshot-delta/README.md

```diff
@@ -0,0 +1,1187 @@
+be used to elect a single leader instance, while idling other non-leader
+instances.
+
+Non-leader instances will voluntarily fail their readiness probe to remove
+themselves from the `Service`'s request path. Optionally, the `Service` can be
+deployed with its `sessionAffinity` set to `clientIP` to ensure paginated
+requests from the same origin are always directed back to the same leader pod.
+
+### API Specification
+
+The section describes the specification of proposed API. The Go types of the CRD
+are defined as follows:
+
+```go
+// VolumeSnapshotDelta is a specification for a VolumeSnapshotDelta resource
+type VolumeSnapshotDelta struct {
+  metav1.TypeMeta   `json:",inline"`
+
+  // +optional
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec   VolumeSnapshotDeltaSpec   `json:"spec"`
+
+  // +optional
+  Status VolumeSnapshotDeltaStatus `json:"status,omitempty"`
+}
+
+// VolumeSnapshotDeltaSpec is the spec for a VolumeSnapshotDelta resource
+type VolumeSnapshotDeltaSpec struct {
+  BaseVolumeSnapshotName   string `json:"baseVolumeSnapshotName,omitempty"` // name of the base VolumeSnapshot; optional
+  TargetVolumeSnapshotName string `json:"targetVolumeSnapshotName"`        // name of the target VolumeSnapshot; required
+  Mode                     string `json:"mode,omitempty"`         // default to "block"
+}
+
+// VolumeSnapshotDeltaStatus is the status for a VolumeSnapshotDelta resource
+type VolumeSnapshotDeltaStatus struct {
+  Error        string `json:"error,omitempty"`
+  State        string `json:"state"`
+  CallbackURL  string `json:"callbackURL"`
+}
+
+// VolumeSnapshotDeltaList is a list of VolumeSnapshotDelta resources
+type VolumeSnapshotDeltaList struct {
+  metav1.TypeMeta          `json:",inline"`
+  metav1.ListMeta          `json:"metadata"`
+  Items []VolumeSnapshotDelta `json:"items"`
+}
+```
+
+The corresponding GRPC service definition is as follows:
+
+```grpc
+service VolumeSnapshotDelta {
+  rpc GetVolumeSnapshotDelta(VolumeSnapshotDeltaRequest)
+    returns (VolumeSnapshotDeltaResponse) {}
+}
+
+type VolumeSnapshotDeltaRequest struct {
+    // If SnapshotBase is not specified, return all used blocks.
+    SnapshotBase       string         // Snapshot handle, optional.
+    SnapshotTarget     string         // Snapshot handle, required.
+    Mode               string         // Volume mode, default "Block"
+    StartOffset        string         // Logical offset from beginning of disk/volume.
+                                      // Use string instead of uint64 to give vendor
+                                      // the flexibility of implementing it either
+                                      // string "token" or a number.
+    MaxEntries         uint64         // Maximum number of entries in the response
+}
+
+type VolumeSnapshotDeltaResponse struct {
+    ChangeBlockList   []ChangedBlock  // array of ChangedBlock (for "Block" mode)
```

Comment:
If you're going to support Filesystem deltas in the future in the same message, then the changed blocks should be inside a separate struct:

```
type VolumeSnapshotDeltaResponse struct {
    BlockDelta *BlockVolumeSnapshotDelta
    NextOffset ...
}

type BlockVolumeSnapshotDelta struct {
    ChangedBlockList []ChangedBlock
}
```

So you can add `MountVolumeSnapshotDelta` in the future and it's clear which field is for block and which is for file. Like `VolumeCapability` has `BlockVolume` and `MountVolume`.